compiler/rustc_metadata/src/rmeta/decoder.rs

```diff
@@ -1100,7 +1101,11 @@ impl<'a, 'tcx> CrateMetadataRef<'a> {
                     let vis = self.get_visibility(child_index);
                     let def_id = self.local_def_id(child_index);
                     let res = Res::Def(kind, def_id);
-                    callback(Export { res, ident, vis, span });
+
+                    if !kind.is_macro() {
```

Comment:
I was going to mention this, and then I forgot. Serves me right for not leaving a comment. The truth of the matter is, I have absolutely no clue why I needed to skip macros here. If I don't do so, I get errors (which it took me five+ hours to debug :/). I think those may come from situations in which a macro is defined as a macro by example, but also defined as a builtin macro. Here's what the errors look like:

```
error[E0428]: the name `Clone` is defined multiple times
   --> /rust/library/core/src/clone.rs:139:1
    |
139 | pub macro Clone($item:item) {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    | |
    | `Clone` redefined here
    | previous definition of the macro `Clone` here
    |
    = note: `Clone` must be defined only once in the macro namespace of this module

error[E0428]: the name `Copy` is defined multiple times
   --> /rust/library/core/src/marker.rs:393:1
    |
393 | pub macro Copy($item:item) {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
    | |
    | `Copy` redefined here
    | previous definition of the macro `Copy` here
    |
    = note: `Copy` must be defined only once in the macro namespace of this module
```

I tried to fix this by not encoding the macros, rather than by skipping decoding them, but I couldn't get that to work either. Any insight into what might be going on here would be helpful. Although this solution works, I have a strong feeling that it likely isn't the right solution.

Reply:
I think I've got at least part of the story. This error is created by resolution code when it enumerates all the children of a given module. Formerly, the macro did not appear in the HIR module list of items (your change to lower_item_ref), so this code path never included any macro. Now, macros are listed in the module, so appear when resolution tries to construct the reduced graph. Except that the resolver already lists all macros as reexports (rustc_resolve::imports::finalize_resolutions_in, line 1396).
Does removing the `|| binding.is_macro_def()` there work?

Reply:
Well, it changes the errors I get. Here's a sample of the new ones:

```
error[E0432]: unresolved import `core::format_args`
   --> library/alloc/src/lib.rs:197:13
    |
197 |     pub use core::format_args;
    |             ^^^^^^^^^^^^^^^^^ no `format_args` in the root

error: cannot find macro `panic` in this scope
   --> library/alloc/src/alloc.rs:387:9
    |
387 |         panic!("memory allocation of {} bytes failed", size)
    |         ^^^^^

error: cannot find macro `unreachable` in this scope
   --> library/alloc/src/borrow.rs:281:37
    |
281 |                     Borrowed(..) => unreachable!(),
    |                                     ^^^^^^^^^^^

error: cannot find macro `write` in this scope
   --> library/alloc/src/collections/btree/map/entry.rs:100:9
    |
100 |         write!(
    |         ^^^^^
```

Reply:
This got resolved [down here](https://github.com/rust-lang/rust/pull/88019#discussion_r691626636). Not sure why that got started as a seperate thread.